Localization test failures 1

MaStr11 opened the pull request:
Fix for #23837.

This is the first of a series of PRs meant to resolve the unit test failures caused by missing localizations.

This PR includes test related to csi.exe and interactive scripting:

### Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests
Done. Tested local (de-DE) and CI (en-US).
### Microsoft.CodeAnalysis.CSharp.Scripting.Desktop.UnitTests
Done. Tested local (de-DE) and CI (en-US).
### Microsoft.CodeAnalysis.VisualBasic.Scripting.UnitTests
Done. Tested local (de-DE) and CI (en-US).
### Roslyn.InteractiveHost.UnitTests 
Done. Tested local (de-DE) and CI (en-US).

MaStr11 commented on src/Scripting/CSharpTest.Desktop/CsiTests.cs:
```diff
@@ -50,19 +51,21 @@ public void CurrentWorkingDirectory_Change()
 > 
-", result.Output);
+";
+            expected = expected.Replace((char)0x2013, (char)0x002d); // EN DASH -> HYPHEN-MINUS
```
csi.exe outputs the [*en dash*](http://www.fileformat.info/info/unicode/char/2013/index.htm) in `Microsoft (R) Visual C# - interaktive Compilerversion 42.42.42.42424` as [*hyphen-minus*](http://www.fileformat.info/info/unicode/char/2d/index.htm). There might be a more reliable way to fix this.

tmeschter replied:
@MaStr11 Please add a more detailed comment explaining why the replacement is necessary. Something like:

> The German translation (and possibly others) contains an en dash (0x2013), but csi.exe outputs it as a hyphen-minus (0x002d). We need to fix up the expected string before we can compare it to the actual output.

sharwell commented on src/Scripting/CSharpTest.Desktop/CsiTests.cs:
```diff
@@ -50,19 +51,21 @@ public void CurrentWorkingDirectory_Change()
-Microsoft (R) Visual C# Interactive Compiler version {s_compilerVersion}
-Copyright (C) Microsoft Corporation. All rights reserved.
+            var expected = $@"
+{ string.Format(CSharpScriptingResources.LogoLine1, s_compilerVersion) }
```
@MaStr11 I should be able to clean this up in the future for all of this repository using automated tools. This feedback is more for future work.

sharwell commented:
> TODO: Uncomment //[UseCulture("en-US")] once #24398 is merged

#24398 is now merged.

sharwell commented:
@MaStr11 Can you update the original comment to replace the TODO text with the current status?

MaStr11 commented:
> #24398 is now merged.

I gave it a try and I can confirm that it indeed works as intended. Ironically it did help in this particular case because the test did a `process.Start` and captured the output from the running `csi.exe`. So the culture of the test itself changes nothing as this always tied to the culture of the machine.

sharwell commented:
Requesting reviews from @dotnet/roslyn-compiler and @dotnet/roslyn-interactive

tmat commented on src/Scripting/CSharpTest/CommandLineRunnerTests.cs:
```diff
@@ -132,14 +132,19 @@ public void Await()
+            var logoOutput = $@"{ string.Format(CSharpScriptingResources.LogoLine1, s_compilerVersion) }
+{ CSharpScriptingResources.LogoLine2}
+
+{ ScriptingResources.HelpPrompt}";
```
Could you extract to a static field (similar to s_compilerVersion) and then use the field in all test cases?

MaStr11 commented:
Sure I could. But there are cases where you want your tests to run in a particular culture using the [UseCulture] attribute. So there is a benefit in keeping the tests self contained.

tmat commented on src/Interactive/HostTest/InteractiveHostTests.cs:
```diff
@@ -986,8 +987,8 @@ public void Script_NoHostNamespaces()
-            AssertEx.AssertEqualToleratingWhitespaceDifferences(@"
-(1,8): error CS0234: The type or namespace name 'CodeAnalysis' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)",
+            AssertEx.AssertEqualToleratingWhitespaceDifferences($@"
+(1,8): error CS0234: { string.Format(CSharpResources.ERR_DottedTypeNameNotFoundInNS, "CodeAnalysis", "Microsoft") }",
```
>  [](start = 22, length = 1)

We do not add spaces around expressions in interpolated strings anywhere in the Roslyn repo. It would be good to keep the style consistent.

tmat commented on src/Scripting/VisualBasicTest/CommandLineRunnerTests.vb:
```diff
@@ -55,10 +55,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Scripting.UnitTests
+VBScriptingResources.LogoLine2 + "
 
-Type ""#help"" for more information.
+" + ScriptingResources.HelpPrompt + "
```
Extract to a static field

tmat commented:
:clock1:

tmat commented:
There are perhaps a few of such tests that use a different culture. They would not use this field then. Majority tests do not care about the "logo" lines and it would make them more readable. Also authoring a new test would be easier.

sharwell commented:
💭 I tend to prefer self-contained tests, and don't see the issue with having the string inlined in multiple tests.

tmat commented:
@sharwell The tests are already not self-contained anyways -- they use `s_compilerVersion` and other helpers.

MaStr11 commented:
@tmat I'm currently changing the tests. I also facing a problem with the tests that change the Culture during the test (It seems the culture is persisted. So half of the tests run in de-DE while the other run in en-GB). I will need some more time to fix this.

MaStr11 commented:
I'm done with the changes now (Sorry for the messy commit history).

sharwell commented on src/Scripting/VisualBasicTest/CommandLineRunnerTests.vb:
```diff
@@ -160,27 +146,33 @@ logoText + "
-
-" + ScriptingResources.HelpPrompt
-            Dim runner = CreateRunner(args:={}, input:="Imports System.Globalization
+            ' Save the current thread culture as it is changed in the test.
```
:bulb: The `UseCulture` attribute should restore the culture at the end of the test. Is it possible to use it here?

MaStr11 replied:
There are two differences.

1. UseCulture sets *Current thread.Culture* while this test sets *CultureInfo.DefaultThreadCulture*.
2. UseCulture also sets the culture at the beginning of the test

I think it is better the way it is but haven't tried UseCulture.

MaStr11 commented:
> The UseCulture attribute should restore the culture at the end of the test. Is it possible to use it here?

@sharwell I gave it a try in commit 8e05074. I changed UseCultureAttribute to also set `DefaultThreadCurrentUICulture`. But doing so caused all kinds of unpredictable side-effects (It seemed to me that the `UseCultureAttribute.After` method is not able to revert to the previous system state by setting `DefaultThreadCurrentUICulture`. I just gave up after a while.) and so I reverted the change.

@tmat I found to problems with vbi.exe during my tests:

1. Inputing `>`and pressing enter gives a NRE in `SyntaxFactory.IsPartOfLinqQueryNotFollowedByNewLine`. Is this a known bug or should I file an issue?
2. Is it expected that `? New System.DateTime` always outputs `#1/1/0001 12:00:00 AM#` no matter what DefaultThreadCurrentUICulture is set?

MaStr11 commented on src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj:
```diff
@@ -86,6 +86,7 @@
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.Scripting.Destkop.UnitTests" />
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests" />
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.CSharp.Scripting.Desktop.UnitTests" />
+    <InternalsVisibleToTest Include="Roslyn.InteractiveHost.UnitTests" />
```
@sharwell Since you asked for approval for *InternalsVisibleToTest* in #24460 you may also want approval for this one.

sharwell replied:
👍 This one should be covered under the umbrella of the review from @jaredpar.

sharwell commented:
@tmat Can you answer the two final questions from @MaStr11 in the last comment above?

AlekseyTs commented:
@sharwell This PR was merged without getting required sign-offs from the compiler team and it doesn't deal with localization issues in the manner that other compiler tests do. I think it should be reverted and different fixes should be made.

sharwell commented:
@AlekseyTs This pull request did not touch compiler code.

AlekseyTs commented:
This PR made changes to CSharpCodeAnalysis.csproj, it also touched Scripting and Csi tests, even though they are not under Compilers folder, they used to be owned by Compiler team at some point. @jaredpar To clarify the current ownership for the tests.
